### ISSUE TITLE
Crash in core library: TypeError int is not a subtype of _FutureListener<dynamic, dynamic>?

### ISSUE THREAD

**Issue by jonahwilliams** [dart-lang/sdk#43662]
While scrubbing through the flutter tool crash logs I've found a few that seem to be bugs in the dart core libraries. I do not have instructions to repro this - it seems fairly rare and not _that_ severe but does seem to indicate that the future implementation has some edge cases.

```
_TypeError: type 'int' is not a subtype of type '_FutureListener<dynamic, dynamic>?'

  | at Future._removeListeners | (future_impl.dart:436)
-- | -- | --
  | at Future._propagateToListeners | (future_impl.dart:759)
  | at Future._completeWithValue | (future_impl.dart:529)
  | at Future._asyncCompleteWithValue.<anonymous closure> | (future_impl.dart:567)
  | at _rootRun | (zone.dart:1190)
  | at _CustomZone.run | (zone.dart:1093)
  | at _CustomZone.runGuarded | (zone.dart:997)
  | at _CustomZone.bindCallbackGuarded.<anonymous closure> | (zone.dart:1037)
  | at _microtaskLoop | (schedule_microtask.dart:41)
  | at _startMicrotaskLoop | (schedule_microtask.dart:50)
  | at _runPendingImmediateCallback | (isolate_patch.dart:118)
  | at _RawReceivePortImpl._handleMessage | (isolate_patch.dart:169)
```

Flutter Version dev/1.23.0-4.0.pre
Dart SDK version: 2.10.0-136.0.dev (dev) (Wed Sep 16 13:30:26 2020 -0700) on "macos_x64"

**Comment by mit-mit** [comment]
cc @lrhn

**Comment by lrhn** [comment]
This looks like something which can happen if code attempts to remove listeners after the future has already been completed.
We use the same field to hold the listeners and the value, and use the state field to remember which mode we are in.
This is triggered by an `_asyncCompleteWithValue`.

If the state logic is correct, this cannot happen. It happens, so we probably have a bug.

**Comment by lrhn** [comment]
Since we haven't been able to reproduce this, I'll close the issue.
If we hear of a similar incident again, I'll know where to find it.

**Comment by blaugold** [comment]
I'm seeing a similar error:

```
type 'Future<BannerAd?>' is not a subtype of type '_FutureListener<dynamic, dynamic>?'

String: type 'Future<BannerAd?>' is not a subtype of type '_FutureListener<dynamic, dynamic>?'
  File "future_impl.dart", line 501, in Future._removeListeners
  File "future_impl.dart", line 647, in Future._completeError
  File "future.dart", line 1333, in _completeWithErrorCallback
  File "future.dart", line 429, in new Future.delayed.<fn>
  File "ticker.dart", line 420, in TickerFuture.whenCompleteOrCancel.thunk
  File "timer_impl.dart", line 398, in _Timer._runTimers
  File "timer_impl.dart", line 429, in _Timer._handleMessage
  File "isolate_patch.dart", line 189, in _RawReceivePort._handleMessage
```

At the same time, I'm seeing this `StackOverflowError`. Maybe it's correlated somehow:

```
StackOverflowError: Stack Overflow
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  File "future_impl.dart", in Future._addListener
  ?, in ...
  ?, in ...
```

**Comment by lrhn** [comment]
The stack overflow does ring a bell. I did a rewrite of `Future` that never landed (yet), but one bug I noticed and fixed was that there was no attempt to prevent a future being completed with itself.
```dart
Completer c = Completer<void>();
c.complete(c.future);
```
This sets up `c.future` to wait for itself. That's handled internally as pointing to itself as its source future, and *forwarding `_addListener` calls to its source future*. Which would precisely give this result.

That also means that `_addListener` throws, likely in the middle of doing something else, which can be why the future gets into an inconsistent state where it thinks it can be completed, even if it can't.
